core/src/main/java/hudson/model/FileFingerprintStorage.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2020, Sumit Sarin and Jenkins project contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.model;
+
+import com.thoughtworks.xstream.converters.basic.DateConverter;
+import hudson.Extension;
+import hudson.Util;
+import hudson.XmlFile;
+import hudson.model.listeners.SaveableListener;
+import hudson.util.AtomicFileWriter;
+import hudson.util.PersistedList;
+import jenkins.model.FingerprintFacet;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Default file system storage implementation for fingerprints.
+ *
+ * @author Sumit Sarin
+ */
+@Extension()
+public class FileFingerprintStorage extends FingerprintStorage {
```

Comment:
It would make sense to at least restrict the API as Beta.
https://www.jenkins.io/blog/2018/04/30/using-the-beta-annotation/

Reply:
@oleg-nenashev https://github.com/jenkinsci/jenkins/pull/4731#discussion_r437370967 was changed accordingly